pyprusalink/client.py

```diff
@@ -0,0 +1,89 @@
+from __future__ import annotations
+
+import hashlib
+import asyncio
+from collections.abc import AsyncGenerator
+from contextlib import asynccontextmanager
+from aiohttp import ClientResponse, ClientSession
+
+from pyprusalink.types import InvalidAuth, Conflict
+
+
+class ApiClient:
+    def __init__(
+        self, session: ClientSession, host: str, username: str, password: str
+    ) -> None:
+        self._session = session
+        self.host = host
+        self._username = username
+        self._password = password
+        self._lock = asyncio.Lock()
+
+        self._realm = ""
+        self._nonce = ""
+        self._qop = ""
+
+    def _generate_headers(self, method: str, path: str) -> dict:
+        ha1 = hashlib.md5(
+            f"{self._username}:{self._realm}:{self._password}".encode()
+        ).hexdigest()
+        ha2 = hashlib.md5(f"{method}:{path}".encode()).hexdigest()
+        response_value = hashlib.md5(f"{ha1}:{self._nonce}:{ha2}".encode()).hexdigest()
+
+        headers = {
+            "Authorization": f'Digest username="{self._username}", realm="{self._realm}", '
+            f'nonce="{self._nonce}", uri="{path}", response="{response_value}"'
+        }
+
+        return headers
+
+    def _extract_digest_params(self, headers: dict[str]) -> None:
+        header_value = headers.get("WWW-Authenticate", "")
+        if not header_value.startswith("Digest"):
+            return
+
+        header_value = header_value[len("Digest ") :]
+
+        params = {}
+        parts = header_value.split(",")
+        for part in parts:
+            key, value = part.strip().split("=", 1)
+            params[key.strip()] = value.strip(' "')
+
+        self._realm = params["realm"]
+        self._nonce = params["nonce"]
+        self._qop = params.get("qop", "auth")
+
+    @asynccontextmanager
+    async def request(
+        self, method: str, path: str, json_data: dict | None = None
+    ) -> AsyncGenerator[ClientResponse, None]:
+        """Make a request to the PrusaLink API."""
+        async with self._lock:
+            url = f"{self.host}{path}"
+            headers = self._generate_headers(method=method, path=path)
+            async with self._session.request(
+                method, url, json=json_data, headers=headers
+            ) as response:
+                if response.status == 401:
+                    self._extract_digest_params(response.headers)
+                    headers = self._generate_headers(method=method, path=path)
+
+                    async with self._session.request(
+                        method, url, json=json_data, headers=headers
+                    ) as refreshed_response:
```

Comment:
Is this retry logic required to work around the session issue you mentioned in the main thread? Is that a problem on particular endpoint or a general issue? :thinking: 

Maybe adding debug logging would be a good idea here :thinking: 

In general, I think this can be done nicer in a loop, with a `retry_on_auth_error = 1`

Reply:
Yeah, it seems to me that the HTTP Digest Auth implementation in the Prusa firmware is broken.

In a perfect world, with each request, we'd get back a new nonce key in the header that we can use for the next request, and we should not be able to reuse a nonce.

What I noticed is that we just get back the same nonce every time, and we can use that multiple times (up to 15 seconds IIRC) before it expires. I also noticed we can just increment the nonce by 1, and it remains valid, but that doesn't change the overall lifetime.

Since Home Assistant pulls data every 30 seconds, it means that we basically have to re-auth every 30 seconds, but we can use that same token for all the API requests that Home Assistant does. (4 requests IIRC)

Reply:
One additional thing: I have no idea how the UI handles this. It seems like they don't have to re-auth every 15 seconds.

Reply:
> What I noticed is that we just get back the same nonce every time, and we can use that multiple times (up to 15 seconds IIRC) before it expires. I also noticed we can just increment the nonce by 1, and it remains valid, but that doesn't change the overall lifetime.

Uh, that seems to be some implementation details/hacks you've uncovered right there :sweat_smile: 

I am fine with the retry, if it is required. But I'd rather make it as a loop. This way error code returns need to get handled once only.

Reply:
Yeah, definitely need to add a comment about that.

I just did some quick digging but still no clue how the frontend handles that. Even looking at the code, it doesn't make any sense: https://github.com/prusa3d/Prusa-Link-Web/blob/1ac7e92600ccdd2d75fbbe79a30e953c342b1e5d/src/auth.js#L85-L86

Maybe the browser does the heavy lifting.

Reply:
> One additional thing: I have no idea how the UI handles this. It seems like they don't have to re-auth every 15 seconds.

From what I can see here is the frontend polls the `/status` endpoint every second, and the digest changes every 5s.

Reply:
In your case, is the problem observable with the main.py from your PR?

Reply:
Seems this is resolved on my MK4 (technically MK3.9, but control should be the same)  :thinking: 

```
$ ./main.py  
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
Version: {'api': '2.0.0', 'server': '2.1.2', 'nozzle_diameter': 0.4, 'text': 'PrusaLink', 'hostname': '', 'capabilities': {'upload-by-put': True}}
```

Reply:
Looks like you didn't add the debug line in `client.py` line 72?

Reply:
> Looks like you didn't add the debug line in `client.py` line 72?

:man_facepalming: 

Yeah, with that, the same result.

Reply:
> I just did some quick digging but still no clue how the frontend handles that. Even looking at the code, it doesn't make any sense: https://github.com/prusa3d/Prusa-Link-Web/blob/1ac7e92600ccdd2d75fbbe79a30e953c342b1e5d/src/auth.js#L85-L86
> 

So I checked this with Wireshark, the frontend handles 401 every 5s just like we have to. It doesn't show up in the debug tools for some reason, but the code you linked does exactly that (on 401 causes a recursive `setUpAuth()`).

> Maybe the browser does the heavy lifting.

Yes, that seems exactly to be the case:

From: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#digest_authentication_with_sha-256_and_md5

> The client prompts the user for their username and password, and then responds with a new request that encodes the credentials in the Authorization header field.

The browser creates the dialog which asks for user/password.